Stephen Munene

0712291940

munesteve@gmail.com

Pima Indians classification assignment

In [36]:
import pandas as pd

pima_data = pd.read_csv('https://modcom.co.ke/data/datasets/pima.csv')
pima_data

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [37]:
pima_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [38]:
pima_data.corr()

Pregnancies   Glucose  BloodPressure  SkinThickness  \
Pregnancies                  1.000000  0.129459       0.141282      -0.081672   
Glucose                      0.129459  1.000000       0.152590       0.057328   
BloodPressure                0.141282  0.152590       1.000000       0.207371   
SkinThickness               -0.081672  0.057328       0.207371       1.000000   
Insulin                     -0.073535  0.331357       0.088933       0.436783   
BMI                          0.017683  0.221071       0.281805       0.392573   
DiabetesPedigreeFunction    -0.033523  0.137337       0.041265       0.183928   
Age                          0.544341  0.263514       0.239528      -0.113970   
Outcome                      0.221898  0.466581       0.065068       0.074752   

                           Insulin       BMI  DiabetesPedigreeFunction  \
Pregnancies              -0.073535  0.017683                 -0.033523   
Glucose                   0.331357  0.221071                  0.137337   
BloodPressure             0.088933  0.281805                  0.041265   
SkinThickness             0.436783  0.392573                  0.183928   
Insulin                   1.000000  0.197859                  0.185071   
BMI                       0.197859  1.000000                  0.140647   
DiabetesPedigreeFunction  0.185071  0.140647                  1.000000   
Age                      -0.042163  0.036242                  0.033561   
Outcome                   0.130548  0.292695                  0.173844   

                               Age   Outcome  
Pregnancies               0.544341  0.221898  
Glucose                   0.263514  0.466581  
BloodPressure             0.239528  0.065068  
SkinThickness            -0.113970  0.074752  
Insulin                  -0.042163  0.130548  
BMI                       0.036242  0.292695  
DiabetesPedigreeFunction  0.033561  0.173844  
Age                       1.000000  0.238356  
Outcome                   0.238356  1.000000

In [39]:
#Get X and Y from the data (Input and Outcome) split the data into 70% training and 30% testing

pv = pima_data.values #Values from the original data
X = pv[:,:8] #All rows 7 cols 
Y = pv[:,8] #all rows last col (outcome)

#dsplit into 70% training 30%testing
from sklearn import model_selection

X_train,X_test,Y_train,Y_test = model_selection.train_test_split(X,Y, test_size = 0.30, random_state=42)

#import several classification models for crossvalidation
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier

#crossvalidate the model
models = []
models.append(('LR', LogisticRegression(solver='liblinear', multi_class='ovr')))
models.append(('LDA', LinearDiscriminantAnalysis()))
models.append(('KNN', KNeighborsClassifier()))
models.append(('DT', DecisionTreeClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC(gamma='auto')))
models.append(('RF', RandomForestClassifier()))
models.append(('GB', GradientBoostingClassifier()))

# evaluate each model in turn
for name, model in models:
      kfold = model_selection.KFold(n_splits=10, random_state=42, shuffle=True)
      cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring='accuracy')
      print(name, cv_results.mean())

LR 0.7654437456324248
LDA 0.7729559748427673
KNN 0.7208944793850453
DT 0.7244933612858141
NB 0.7507686932215234
SVM 0.6498602375960866
RF 0.7728860936408106
GB 0.7691125087351502


Best performing model appears to be LinearDiscriminantAnalysis with a 77.3% accuracy

In [40]:
#proceed to fit the best performong model from the ones tested 

model = LinearDiscriminantAnalysis()
model.fit(X_train,Y_train)

#Test the model
predictions = model.predict(X_test)

#overview on performance

print('\t \t Model Predictions \n',predictions)

print('\n \t\t Expected results \n', Y_test)

	 	 Model Predictions 
 [0. 0. 0. 0. 1. 0. 0. 1. 1. 1. 0. 1. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0.
 1. 1. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 0. 1. 1. 0. 0. 1. 1. 0.
 0. 1. 0. 1. 1. 0. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 0. 1. 0. 1. 0. 1. 1. 0.
 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0. 0. 0. 1. 0. 1.
 0. 0. 1. 0. 1. 0. 1. 1. 1. 0. 0. 1. 0. 0. 0. 0. 0. 1. 0. 0. 1. 0. 0. 1.
 0. 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 0. 0.
 1. 1. 1. 0. 0. 1. 1. 0. 0. 0. 0. 0. 1. 1. 0. 1. 1. 0. 0. 0. 1. 0. 0. 0.
 1. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 1. 1. 0. 0. 0. 0. 1. 0.
 1. 0. 0. 1. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]

 		 Expected results 
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 0.
 0. 1. 0. 0. 1. 0. 1. 1. 1. 1. 0. 1. 1. 1. 0. 1. 0. 0. 0. 1. 0. 1. 1. 0.
 0. 0. 0. 1. 1. 1. 0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 0. 1. 0. 0.
 0. 1. 0. 0. 0. 0. 0. 0. 1. 1

In [41]:
#Deeper look at the model performance

from sklearn.metrics import accuracy_score
print('Model Test accuracy', accuracy_score(Y_test,predictions))

Model Test accuracy 0.7316017316017316


Model got a 73.1% accuracy score with the Test data compared to the 77.3% accuracy score in training.
Taking a deeper look

In [42]:
# report on the model performance
from sklearn.metrics import classification_report
print(classification_report(Y_test,predictions))

              precision    recall  f1-score   support

         0.0       0.79      0.79      0.79       151
         1.0       0.61      0.61      0.61        80

    accuracy                           0.73       231
   macro avg       0.70      0.70      0.70       231
weighted avg       0.73      0.73      0.73       231



From the report above, it can be noted that LinearDiscriminantAnalysis got a 79% precision score on negative tests and a 61% precision score on the positive score. which give the weighted average accuracy score of 73% observed before.

In [43]:
#report on where the model went wrong the actual numbers on how many were right or wrong.
from sklearn.metrics import confusion_matrix
print( confusion_matrix(Y_test,predictions))

[[120  31]
 [ 31  49]]


From the above confusion matrix, it can be noted that LinearDiscriminantAnalysis on the negative outcome got 120 out of the possible 151 predictions right, and got 31 wrong giving us a 79% precision score.
However, on the positive outcome got 61 out of the possible 80 right, and git 31 wrong giving us a 61% accuracy score.


In [44]:
#testing out the second best performing model the RandomForestClassifier
rf_model = RandomForestClassifier()

rf_model.fit(X_train,Y_train)

#testing out the model 
rf_predictions = model.predict(X_test)

#overview on performance

print('\t \t Model Predictions \n',rf_predictions)

print('\n \t\t Expected results \n', Y_test)

	 	 Model Predictions 
 [0. 0. 0. 0. 1. 0. 0. 1. 1. 1. 0. 1. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0.
 1. 1. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 0. 1. 1. 0. 0. 1. 1. 0.
 0. 1. 0. 1. 1. 0. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 0. 1. 0. 1. 0. 1. 1. 0.
 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0. 0. 0. 1. 0. 1.
 0. 0. 1. 0. 1. 0. 1. 1. 1. 0. 0. 1. 0. 0. 0. 0. 0. 1. 0. 0. 1. 0. 0. 1.
 0. 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 0. 0.
 1. 1. 1. 0. 0. 1. 1. 0. 0. 0. 0. 0. 1. 1. 0. 1. 1. 0. 0. 0. 1. 0. 0. 0.
 1. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 1. 1. 0. 0. 0. 0. 1. 0.
 1. 0. 0. 1. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]

 		 Expected results 
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 0.
 0. 1. 0. 0. 1. 0. 1. 1. 1. 1. 0. 1. 1. 1. 0. 1. 0. 0. 0. 1. 0. 1. 1. 0.
 0. 0. 0. 1. 1. 1. 0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 0. 1. 0. 0.
 0. 1. 0. 0. 0. 0. 0. 0. 1. 1

In [45]:
#Deeper look at the model performance

from sklearn.metrics import accuracy_score
print('Model Test accuracy', accuracy_score(Y_test,rf_predictions))

Model Test accuracy 0.7316017316017316


Model got a 73.1% accuracy score with the Test data compared to the 77.1% accuracy score in training.
Taking a deeper look

In [46]:
# report on the model performance
from sklearn.metrics import classification_report
print(classification_report(Y_test,rf_predictions))

              precision    recall  f1-score   support

         0.0       0.79      0.79      0.79       151
         1.0       0.61      0.61      0.61        80

    accuracy                           0.73       231
   macro avg       0.70      0.70      0.70       231
weighted avg       0.73      0.73      0.73       231



From the report above, it can be noted that RandomForestClassifier got a 79% precision score on negative tests and a 61% precision score on the positive score. which give the weighted average accuracy score of 73% observed before.


In [47]:
#report on where the model went wrong the actual numbers on how many were right or wrong.
from sklearn.metrics import confusion_matrix
print( confusion_matrix(Y_test,rf_predictions))

[[120  31]
 [ 31  49]]


From the above confusion matrix, it can be noted that RandomForestClassifier on the negative outcome got 120 out of the possible 151 predictions right, and got 31 wrong giving us a 79% precision score. However, on the positive outcome got 61 out of the possible 80 right, and git 31 wrong giving us a 61% accuracy score.


##### Performance of the two models was quite similar tetsing them out on new data

In [48]:
newdata = [[4,120,69,20,79,31,0.47,33],[3,117,72,23,30,32,0.37,29],[17,199,122,99,846,67,2.42,81],[6,140,80,32,127,36,0.62,41]]

new_lda = model.predict(newdata)
print('LinearDiscriminantAnalysis predictions on the new data is likely \t', new_lda, '\n')
new_rf = rf_model.predict(newdata)
print('RandomForestClassifier predictions on the new data is likely \t\t', new_rf)

LinearDiscriminantAnalysis predictions on the new data is likely 	 [0. 0. 1. 1.] 

RandomForestClassifier predictions on the new data is likely 		 [0. 0. 1. 1.]


From the newdata presented, we can see that both models yielded the same predicions on the 4 sets giving a result of [0,0,1,1].
Two negatives and two positives